### PR TITLE
App defaults to requiring all children up

### PR DESCRIPTION
**Pull request by aledsage**

### CODE REVIEW

**Comment by neykov** [comment]
A couple of failing tests:
```
Tests run: 1931, Failures: 2, Errors: 0, Skipped: 0, Time elapsed: 235.556 sec <<< FAILURE! - in TestSuite
testOnlyChildFailsToStartThenRecoversCausesAppToRecover(org.apache.brooklyn.core.entity.ApplicationLifecycleStateTest)  Time elapsed: 30.283 sec  <<< FAILURE!
org.apache.brooklyn.util.exceptions.PropagatedRuntimeException: failed succeeds-eventually, 69 attempts, 30001ms elapsed: AssertionError: entity=Application[xof1rmsy]; attribute=Sensor: service.state (org.apache.brooklyn.core.entity.lifecycle.Lifecycle) expected [running] but found [on-fire]
	at org.apache.brooklyn.test.Asserts.fail(Asserts.java:726)
	at org.apache.brooklyn.test.Asserts.failNotEquals(Asserts.java:116)
	at org.apache.brooklyn.test.Asserts.assertEquals(Asserts.java:438)
	at org.apache.brooklyn.core.entity.EntityAsserts.assertAttributeEquals(EntityAsserts.java:56)
	at org.apache.brooklyn.core.entity.EntityAsserts$1.run(EntityAsserts.java:72)
	at org.apache.brooklyn.test.Asserts$RunnableAdapter.call(Asserts.java:1320)
	at org.apache.brooklyn.test.Asserts.succeedsEventually(Asserts.java:939)
	at org.apache.brooklyn.test.Asserts.succeedsEventually(Asserts.java:863)
	at org.apache.brooklyn.core.entity.EntityAsserts.assertAttributeEqualsEventually(EntityAsserts.java:69)
	at org.apache.brooklyn.core.entity.EntityAsserts.assertAttributeEqualsEventually(EntityAsserts.java:64)
	at org.apache.brooklyn.core.entity.ApplicationLifecycleStateTest.assertHealthEventually(ApplicationLifecycleStateTest.java:171)
	at org.apache.brooklyn.core.entity.ApplicationLifecycleStateTest.testOnlyChildFailsToStartThenRecoversCausesAppToRecover(ApplicationLifecycleStateTest.java:79)

testSomeChildFailsToStartThenRecoversCausesAppToRecover(org.apache.brooklyn.core.entity.ApplicationLifecycleStateTest)  Time elapsed: 30.26 sec  <<< FAILURE!
org.apache.brooklyn.util.exceptions.PropagatedRuntimeException: failed succeeds-eventually, 69 attempts, 30000ms elapsed: AssertionError: entity=Application[xjmypuja]; attribute=Sensor: service.state (org.apache.brooklyn.core.entity.lifecycle.Lifecycle) expected [running] but found [on-fire]
	at org.apache.brooklyn.test.Asserts.fail(Asserts.java:726)
	at org.apache.brooklyn.test.Asserts.failNotEquals(Asserts.java:116)
	at org.apache.brooklyn.test.Asserts.assertEquals(Asserts.java:438)
	at org.apache.brooklyn.core.entity.EntityAsserts.assertAttributeEquals(EntityAsserts.java:56)
	at org.apache.brooklyn.core.entity.EntityAsserts$1.run(EntityAsserts.java:72)
	at org.apache.brooklyn.test.Asserts$RunnableAdapter.call(Asserts.java:1320)
	at org.apache.brooklyn.test.Asserts.succeedsEventually(Asserts.java:939)
	at org.apache.brooklyn.test.Asserts.succeedsEventually(Asserts.java:863)
	at org.apache.brooklyn.core.entity.EntityAsserts.assertAttributeEqualsEventually(EntityAsserts.java:69)
	at org.apache.brooklyn.core.entity.EntityAsserts.assertAttributeEqualsEventually(EntityAsserts.java:64)
	at org.apache.brooklyn.core.entity.ApplicationLifecycleStateTest.assertHealthEventually(ApplicationLifecycleStateTest.java:171)
	at org.apache.brooklyn.core.entity.ApplicationLifecycleStateTest.testSomeChildFailsToStartThenRecoversCausesAppToRecover(ApplicationLifecycleStateTest.java:95)

2016-11-04 21:25:40,057 INFO  Brooklyn shutdown: stopping entities [TestEntityImpl{id=l7anphjy3k}, TestEntityImpl{id=civa7tpffq}, BlockingEntityImpl{id=hap37gs8zx}]

Results :

Failed tests: 
  ApplicationLifecycleStateTest.testOnlyChildFailsToStartThenRecoversCausesAppToRecover:79->assertHealthEventually:171 » PropagatedRuntime
  ApplicationLifecycleStateTest.testSomeChildFailsToStartThenRecoversCausesAppToRecover:95->assertHealthEventually:171 » PropagatedRuntime

```

**Comment by neykov** [comment]
@aledsage Can you start another build - last one failed and is not accessible now.

**Comment by aledsage** [comment]
@neykov @sjcorbett can one of you review this please?

File: camp/camp-brooklyn/src/test/java/org/apache/brooklyn/camp/brooklyn/catalog/SpecParameterUnwrappingTest.java [hunk 1]
```diff
@@ -61,6 +61,11 @@
 import org.apache.brooklyn.entity.stock.BasicStartable;
 
 public class SpecParameterUnwrappingTest extends AbstractYamlTest {
+    
+    // Expect app to have the following config keys already: 
+    // "application.stop.shouldDestroy", "defaultDisplayName", "quorum.running", "quorum.up", "start.latch"
+    public static final int NUM_APP_DEFAULT_CONFIG_KEYS = 5;
```
Comment (neykov) [inline camp/camp-brooklyn/src/test/java/org/apache/brooklyn/camp/brooklyn/catalog/SpecParameterUnwrappingTest.java]: One thing  adding config keys on the app causes is that they get set on the top-level entity when unwrapping. As you discovered, having to add this constant. Shouldn't cause problems but could be surprising in some cases. Just to keep in mind - nothing to do on this PR.

**Comment by neykov** [comment]
LGTM

**Comment by neykov** [comment]
Saw a failure of `ApplicationLifecycleStateTest.testChildFailuresOnStartButWithQuorumCausesAppToSucceed` in the [windows build](https://builds.apache.org/view/Brooklyn/job/brooklyn-master-windows/112/console)

**Comment by neykov** [comment]
Another [failure](https://builds.apache.org/job/brooklyn-server-master/344/consoleFull):
```
2016-11-17 16:53:33,572 INFO  TESTNG FAILED: "Surefire test" - org.apache.brooklyn.core.entity.ApplicationLifecycleStateTest.testStartsThenChildFailsButWithQuorumCausesAppToSucceed() finished in 133 ms
java.lang.AssertionError: entity=Application[rjkw3sil]; attribute=Sensor: service.state (org.apache.brooklyn.core.entity.lifecycle.Lifecycle) expected [running] but found [on-fire]
	at org.apache.brooklyn.test.Asserts.fail(Asserts.java:754)
	at org.apache.brooklyn.test.Asserts.failNotEquals(Asserts.java:144)
	at org.apache.brooklyn.test.Asserts.assertEquals(Asserts.java:466)
	at org.apache.brooklyn.core.entity.EntityAsserts.assertAttributeEquals(EntityAsserts.java:56)
	at org.apache.brooklyn.core.entity.EntityAsserts$4.run(EntityAsserts.java:125)
	at org.apache.brooklyn.test.Asserts$RunnableAdapter.call(Asserts.java:1351)
	at org.apache.brooklyn.test.Asserts.succeedsContinually(Asserts.java:1034)
	at org.apache.brooklyn.test.Asserts.succeedsContinually(Asserts.java:1013)
	at org.apache.brooklyn.core.entity.EntityAsserts.assertAttributeEqualsContinually(EntityAsserts.java:123)
	at org.apache.brooklyn.core.entity.ApplicationLifecycleStateTest.assertHealthContinually(ApplicationLifecycleStateTest.java:187)
	at org.apache.brooklyn.core.entity.ApplicationLifecycleStateTest.testStartsThenChildFailsButWithQuorumCausesAppToSucceed(ApplicationLifecycleStateTest.java:161)
```

**Pull request merged**